File: April 2025/08.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
using namespace std;

class Solution {
  public:
    void dfs(int node, unordered_set<int> &visited, unordered_map<int, vector<int>> &graph) {
        visited.insert(node);
        for (int neighbor : graph[node]) {
            if (visited.find(neighbor) == visited.end()) {
                dfs(neighbor, visited, graph);
            }
        }
    }

    int countComponents(int V, unordered_map<int, vector<int>> &graph) {
        unordered_set<int> visited;
        int components = 0;
        for (int i = 0; i < V; ++i) {
            if (visited.find(i) == visited.end()) {
                ++components;
                dfs(i, visited, graph);
            }
        }
        return components;
    }

    bool isBridge(int V, vector<vector<int>> &edges, int c, int d) {
        // Build the graph
        unordered_map<int, vector<int>> graph;
        for (const auto &edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        // Count components in the original graph
        int originalComponents = countComponents(V, graph);

        // Remove the edge (c, d)
        graph[c].erase(remove(graph[c].begin(), graph[c].end(), d), graph[c].end());
        graph[d].erase(remove(graph[d].begin(), graph[d].end(), c), graph[d].end());

        // Count components after removing the edge
        int newComponents = countComponents(V, graph);

        // If the number of components increases, it's a bridge
        return newComponents > originalComponents;
    }
};

File: April 2025/08.py
class Solution:
    def isBridge(self, V, edges, c, d):
        from collections import defaultdict

        # Function to create the adjacency list
        def create_graph(V, edges):
            graph = defaultdict(list)
            for u, v in edges:
                graph[u].append(v)
                graph[v].append(u)
            return graph

        # DFS to count components
        def dfs(node, visited, graph):
            visited[node] = True
            for neighbor in graph[node]:
                if not visited[neighbor]:
                    dfs(neighbor, visited, graph)

        # Count connected components
        def count_components(graph):
            visited = [False] * V
            components = 0
            for i in range(V):
                if not visited[i]:
                    components += 1
                    dfs(i, visited, graph)
            return components

        # Create the graph
        graph = create_graph(V, edges)

        # Count components in the original graph
        original_components = count_components(graph)

        # Remove the edge (c, d) from the graph
        graph[c].remove(d)
        graph[d].remove(c)

        # Count components after removing the edge
        new_components = count_components(graph)

        # If the number of components increases, it's a bridge
        return new_components > original_components
